iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/BaseHiveIcebergMetaHook.java

```diff
@@ -0,0 +1,415 @@
+
+    // Remove null values from hms table properties
+    hmsParams.entrySet().removeIf(e -> e.getKey() == null || e.getValue() == null);
+
+    // Remove creation related properties
+    PARAMETERS_TO_REMOVE.forEach(hmsParams::remove);
+
+    setWriteModeDefaults(null, hmsParams, null);
+  }
+
+  Schema schema(Properties properties, org.apache.hadoop.hive.metastore.api.Table hmsTable,
+                        Set<String> identifierFields) {
+    boolean autoConversion = conf.getBoolean(InputFormatConfig.SCHEMA_AUTO_CONVERSION, false);
+
+    if (properties.getProperty(InputFormatConfig.TABLE_SCHEMA) != null) {
+      return SchemaParser.fromJson(properties.getProperty(InputFormatConfig.TABLE_SCHEMA));
+    }
+    List<FieldSchema> cols = Lists.newArrayList(hmsTable.getSd().getCols());
+    if (hmsTable.isSetPartitionKeys() && !hmsTable.getPartitionKeys().isEmpty()) {
+      cols.addAll(hmsTable.getPartitionKeys());
+    }
+    Schema schema = HiveSchemaUtil.convert(cols, autoConversion);
+
+    return getSchemaWithIdentifierFields(schema, identifierFields);
+  }
+
+  private Schema getSchemaWithIdentifierFields(Schema schema, Set<String> identifierFields) {
+    if (identifierFields == null || identifierFields.isEmpty()) {
+      return schema;
+    }
+    Set<Integer> identifierFieldIds = identifierFields.stream()
+            .map(column -> {
+              Types.NestedField field = schema.findField(column);
+              Preconditions.checkNotNull(field,
+                      "Cannot find identifier field ID for the column %s in schema %s", column, schema);
+              return field.fieldId();
+            })
+            .collect(Collectors.toSet());
+
+    List<Types.NestedField> cols = schema.columns().stream()
+            .map(column -> identifierFieldIds.contains(column.fieldId()) ? column.asRequired() : column)
+            .collect(Collectors.toList());
+
+    return new Schema(cols, identifierFieldIds);
+  }
+
+  static PartitionSpec spec(Configuration configuration, Schema schema,
+      org.apache.hadoop.hive.metastore.api.Table hmsTable) {
+
+    Preconditions.checkArgument(!hmsTable.isSetPartitionKeys() || hmsTable.getPartitionKeys().isEmpty(),
+        "We can only handle non-partitioned Hive tables. The Iceberg schema should be in " +
+            InputFormatConfig.PARTITION_SPEC + " or already converted to a partition transform ");
+
+    PartitionSpec spec = IcebergTableUtil.spec(configuration, schema);
+    if (spec != null) {
+      Preconditions.checkArgument(hmsTable.getParameters().get(InputFormatConfig.PARTITION_SPEC) == null,
+          "Provide only one of the following: Hive partition transform specification, or the " +
+              InputFormatConfig.PARTITION_SPEC + " property");
+      return spec;
+    }
+
+    return HMSTablePropertyHelper.getPartitionSpec(hmsTable.getParameters(), schema);
+  }
+
+  void setOrcOnlyFilesParam(org.apache.hadoop.hive.metastore.api.Table hmsTable) {
```

Comment:
should it be protected?

Reply:
I think currently it is not necessary, but added.